flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/SingleInputGate.java

```diff
@@ -185,22 +181,24 @@
 
 	private final Counter numBytesIn;
 
+	private final Runnable closeListener;
+
 	public SingleInputGate(
 		String owningTaskName,
-		JobID jobId,
 		IntermediateDataSetID consumedResultId,
 		final ResultPartitionType consumedPartitionType,
 		int consumedSubpartitionIndex,
 		int numberOfInputChannels,
 		TaskActions taskActions,
 		Counter numBytesIn,
-		boolean isCreditBased) {
+		boolean isCreditBased,
+		Runnable closeListener) {
```

Comment:
For the close listener we could maybe apply a similar trick as with the partition request retriggering. We could for example add a termination future to the `SingleInputGate` which is completed once the gate gets closed. On this future we can register the removal from the `inputGatesById` map. WDYT?